### ISSUE TITLE
Rewrite highlighting with tree-sitter only

### ISSUE THREAD

**Issue by phaazon** [phaazon/kak-tree-sitter#25]
Currently, `kak-tree-sitter` uses `tree-sitter-highlight` to get the capture group for us. This was convenient for the first implementation but:

- The API is pretty poor for a general tree-sitter experience. Especially, now that highlighting works great, I would like to support semantics objects, and I can’t, because `tree-sitter-highlight` doesn’t expose the parsed `Tree`. So it would require a second parse of the buffer, which is a total waste of resources and time.
- It doesn’t support textedit objects (i.e. delta edition). For a very big buffer, editing only a few characters at a few different places will generate something akin to _O(n²)_ diffing, which is going to hurt. See https://github.com/phaazon/kak-tree-sitter/discussions/24.

**Comment by jdugan6240** [comment]
I have a PR open that implements the first bullet point at least: https://github.com/phaazon/kak-tree-sitter/pull/151

I do plan to work on the second bullet point at some point, but one step at a time :)

**Comment by phaazon** [comment]
Thanks for contributing. As I mentioned in your PR, this is not the direction I want the project to go to. We do not want to vendor dependencies, and design updates / choices should be made carefully to make it easier to add new things. `Highlighter` should only be about highlighting, not semantic selections, etc.

I know closing this current issue would be great because we would have (many) more features, but we’re going to have to go with a full `tree-sitter` approach.

**Comment by jdugan6240** [comment]
Fair enough. The architecture of that crate would have made it difficult to expose the necessary state to make editing the parse tree possible, so that was probably a dead end regardless.

TBH I'm probably going to keep working on [my implementation](https://github.com/jdugan6240/kakoune-treesitter), as we seem to have slightly different philosophies on how this should function, judging by the Design Suggestions thread (which is totally fine!), plus I'm not the biggest fan of Rust in general. I'm happy to continue offering suggestions, though.

**Comment by phaazon** [comment]
The work in #212 is probably going to fix this current issue. I want to take the time to review the design to be sure it will be easy for me (or others) to maintain it, but from my first review, it looks solid.
